test/Dapr.E2E.Test/Actors/E2ETests.ReentrantTests.cs

```diff
@@ -0,0 +1,42 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+namespace Dapr.E2E.Test
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Dapr.Actors;
+    using Dapr.E2E.Test.Actors.Reentrancy;
+    using Xunit;
+
+    public partial class E2ETests : IAsyncLifetime
+    {
+        private static readonly int NUM_CALLS = 10;
+
+        [Fact]
+        public async Task ActorCanPerformReentrantCalls()
+        {
+            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(60));
+            var proxy = this.ProxyFactory.CreateActorProxy<IReentrantActor>(ActorId.CreateRandom(), "ReentrantActor");
+
+            await WaitForActorRuntimeAsync(proxy, cts.Token);
+
+            await proxy.ReentrantCall(new ReentrantCallOptions(){ CallsRemaining = NUM_CALLS, });
+            var state = await proxy.GetState();
+            var records = state.Records;
+
+            Assert.True(records.Count == NUM_CALLS * 2);
+            for (int i = 0; i < NUM_CALLS * 2; i++)
+            {
+                Assert.Equal(i < NUM_CALLS, records[i].IsEnter);
+
+                if (i > 0)
+                {
+                    Assert.True(records[i].Timestamp > records[i - 1].Timestamp);
```

Comment:
I'm a little worried about precision here and whether this is good enough resolution everywhere the tests will run (clock frequency varies). Would it be better to use an incrementing atomic inside the test actor?

Reply:
I can definitely switch to the AtomicInt, but isn't this also the point of actors? Since it's the same actor host and the execution is guaranteed single threaded, we should be OK, right? Or are you worried that it'll record the same time, not be out of order?